velero-plugin-for-aws/helpers.go

```diff
@@ -18,28 +18,162 @@ package main
 
 import (
 	"context"
+	"crypto/tls"
+	"net/http"
 	"net/url"
+	"strconv"
+	"strings"
 
 	"github.com/aws/aws-sdk-go/aws/endpoints"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/s3/s3manager"
 	"github.com/pkg/errors"
 )
 
-// GetBucketRegion returns the AWS region that a bucket is in, or an error
-// if the region cannot be determined.
-func GetBucketRegion(bucket string) (string, error) {
-	var region string
+// S3Config holds all the configuration params passed
+type S3Config struct {
+	region                   string
+	s3URL                    string
+	publicURL                string
+	kmsKeyID                 string
+	s3ForcePathStyleVal      string
+	signatureVersion         string
+	credentialProfile        string
+	serverSideEncryption     string
+	insecureSkipTLSVerifyVal string
+	bucket                   string
+	caCert                   string
+	session                  *session.Session
+	publicSession            *session.Session
+}
+
+// NewS3Config returns an instance of the config loaded from by the plugin framework
+func NewS3Config(cfg map[string]string) *S3Config {
+	return &S3Config{
+		region:                   cfg[regionKey],
+		s3URL:                    cfg[s3URLKey],
+		publicURL:                cfg[publicURLKey],
+		kmsKeyID:                 cfg[kmsKeyIDKey],
+		s3ForcePathStyleVal:      cfg[s3ForcePathStyleKey],
+		signatureVersion:         cfg[signatureVersionKey],
+		credentialProfile:        cfg[credentialProfileKey],
+		serverSideEncryption:     cfg[serverSideEncryptionKey],
+		insecureSkipTLSVerifyVal: cfg[insecureSkipTLSVerifyKey],
+		bucket:                   cfg[bucketKey],
+		caCert:                   cfg[caCertKey],
+	}
+}
+
+// Init initializes the S3 config parameters which needs to be fetched/prepared based on the passed options
+func (h *S3Config) Init() error {
+	// AWS (not an alternate S3-compatible API) and region not
+	// explicitly specified: determine the bucket's region
+	var (
+		s3ForcePathStyle      bool
+		insecureSkipTLSVerify bool
+		err                   error
+	)
+	if h.s3ForcePathStyleVal != "" {
+		if s3ForcePathStyle, err = strconv.ParseBool(h.s3ForcePathStyleVal); err != nil {
+			return errors.Wrapf(err, "could not parse %s (expected bool)", s3ForcePathStyleKey)
+		}
+	}
+
+	// prepares the AWS Sessions with the session-config having only known parameters.
+	// Any unknown or derived params will be added further to the initialization
+	if err = h.initAWSSessions(s3ForcePathStyle); err != nil {
+		return errors.Wrap(err, "could not initialize AWS session")
+	}
+
+	// add region to the session config
+	if h.s3URL == "" && h.region == "" {
```

Comment:
This is funky.  h.getBucketRegion gets called once, the structure gets changed and then it should not be called again.  getBucketRegion does not need to be a method, just keep it a func and pass in session and bucket.

Reply:
Got it. Will make that change

Reply:
Done